### PR TITLE
ci: skip builds for documentation-only changes

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented

Documentation-only and repository-template changes currently run the complete dependency, cross-platform build, sanitizer, and functional-test matrix even though they cannot affect the compiled software. For example, the documentation-only [PR #7575](https://github.com/dashpay/dash/pull/7575) consumed approximately 212 runner-minutes in the main CI workflow.

## What was done?

- Collect the complete changed-path set for pull requests, including both sides of renamed files, and force full CI for removals or renames that could invalidate packaging manifests.
- Compare pushes with the last successful CI revision so a canceled source build cannot be followed by a documentation-only skip.
- Classify Markdown, static documentation, and repository templates as build/test exclusion zones while retaining Windows installer inputs.
- Fail open and run the full matrix for unknown paths, empty or incomplete results, API limits, tag pushes, and classifier failures.
- Gate the full build container and dependency-source cache so their downstream dependency, build, and test jobs skip transitively.
- Keep the slim container and full lint job active for excluded changes.
- Escape untrusted filenames before rendering them in the Markdown job summary.
- Add focused unit tests for exclusion boundaries, mixed changes, rename safety, summary escaping, malformed input, and fail-open behavior.

## How Has This Been Tested?

- `python3 .github/workflows/test_classify_ci_changes.py`
- `python3 .github/workflows/test_select_dynamic_runner.py`
- `python3 test/lint/lint-whitespace.py`
- `python3 test/lint/lint-files.py`
- `python3 test/lint/lint-python-utf8-encoding.py`
- `actionlint .github/workflows/build.yml`
- `git diff --check`

A Dash Core source build was not run because this change does not modify source files, headers, dependencies, or the Autotools/CMake build configuration.

## Breaking Changes

None.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone

This pull request was created by Codex.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -39,18 +41,87 @@ jobs:
...
+              } else if (context.eventName === 'push') {
+                const before = context.payload.before;
+                const after = context.payload.after;
+                if (context.ref.startsWith('refs/tags/') || !before || /^0+$/.test(before)) {
+                  complete = false;
+                } else {
+                  const comparison = await github.rest.repos.compareCommitsWithBasehead({
+                    owner: context.repo.owner,
+                    repo: context.repo.repo,
+                    basehead: `${before}...${after}`,
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply .github/workflows/build.yml]: Confirmed and fixed in 2e2e88d76aab. Push classification now compares the current revision with the last successful CI workflow run on the same branch, so the replacement run includes changes from any canceled in-flight run. A missing prior successful run or API error fails open to the full matrix.

---
🤖 Posted autonomously by Codex on behalf of pasta.

File: .github/workflows/classify_ci_changes.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+def is_excluded_path(path: str) -> bool:
+    if not path:
+        return False
+    if path in EXCLUDED_EXACT_PATHS or path.startswith(EXCLUDED_PREFIXES):
+        return True
+
+    suffix = PurePosixPath(path).suffix
+    if suffix == ".md":
+        return True
+    return path.startswith("doc/") and suffix in EXCLUDED_DOC_SUFFIXES
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply .github/workflows/classify_ci_changes.py:32]: Confirmed and fixed in 2e2e88d76aab. Both Windows installer inputs, `COPYING` and `doc/README_windows.txt`, are now explicitly build-relevant, with focused classifier coverage.

---
🤖 Posted autonomously by Codex on behalf of pasta.

File: .github/workflows/classify_ci_changes.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            "- Full build and test matrix: {}\n".format(
+                "required" if result["run_build_tests"] else "skipped"
+            )
+        )
+        file.write("- Decision: `{}`\n".format(result["reason"]))
+        if triggering_paths:
+            file.write("- Triggering paths (up to 20):\n")
+            for changed_path in triggering_paths[:20]:
+                file.write(
+                    "  - `{}`\n".format(json.dumps(changed_path, ensure_ascii=True))
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply .github/workflows/classify_ci_changes.py]: Confirmed and fixed in 5cde031e77fa. Changed paths remain JSON-encoded and any backticks are replaced with the literal `\u0060` escape before the path is placed inside a Markdown code span. A focused regression test covers a filename containing both a backtick and link syntax.

---
🤖 Posted autonomously by Codex on behalf of pasta.

File: .github/workflows/classify_ci_changes.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+
+EXCLUDED_EXACT_PATHS = {
+    "doc/.gitignore",
+}
+EXCLUDED_PREFIXES = (
+    ".github/ISSUE_TEMPLATE/",
+    ".github/PULL_REQUEST_TEMPLATE/",
+)
+EXCLUDED_DOC_SUFFIXES = {".1", ".5", ".8", ".png", ".svg", ".txt"}
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply .github/workflows/classify_ci_changes.py:19]: Confirmed and fixed in 5cde031e77fa at the broader packaging boundary. Any removed or renamed path now forces the full build/test matrix, so stale `dist_man1_MANS` entries and equivalent distribution-manifest failures are exercised while content-only manpage edits can remain in the documentation exclusion zone. Focused classifier coverage verifies the fail-open decision.

---
🤖 Posted autonomously by Codex on behalf of pasta.

File: .github/workflows/test_classify_ci_changes.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
+import importlib.util
+import json
+import pathlib
+import tempfile
+import unittest
+
+
+MODULE_PATH = pathlib.Path(__file__).with_name("classify_ci_changes.py")
```
Comment (knst) [inline .github/workflows/test_classify_ci_changes.py:8]: once again ; tests for CI seems a bit excessive

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -8,6 +8,7 @@ permissions:
   actions: read
   contents: read
   packages: write
+  pull-requests: read
```
Comment (knst) [inline .github/workflows/build.yml:11]: can the whole PR be reduced to just
```
diff --git a/.github/workflows/build.yml b/.github/workflows/build.yml
index ce8191d..fbc1a50 100644
--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -8,6 +8,7 @@ permissions:
   actions: read
   contents: read
   packages: write
+  pull-requests: read
 
 concurrency:
   group: |
@@ -25,6 +26,7 @@ jobs:
     runs-on: ${{ vars.RUNNER_CHECK_SKIP || 'ubuntu-24.04-arm' }}
     outputs:
       skip: ${{ steps.skip-check.outputs.skip }}
+      run-build-tests: ${{ steps.classify-changes.outputs.run-build-tests }}
       runner-amd64: ${{ steps.select-runner.outputs.runner_amd64 }}
       runner-arm64: ${{ steps.select-runner.outputs.runner_arm64 }}
       use-blacksmith: ${{ steps.select-runner.outputs.use_blacksmith }}
@@ -47,6 +49,24 @@ jobs:
             echo "skip=false" >> $GITHUB_OUTPUT
           fi
 
+      - name: Classify changed paths
+        id: classify-changes
+        if: ${{ steps.skip-check.outputs.skip == 'false' }}
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          RUN=true
+          if [[ "${{ github.event_name }}" == "pull_request_target" ]]; then
+            FILES="$(gh api --paginate \
+              "repos/${{ github.repository }}/pulls/${{ github.event.pull_request.number }}/files" \
+              --jq '.[] | .filename, (.previous_filename // empty)')" || FILES=""
+            if [[ -n "$FILES" ]] && ! grep -qvE '\.md$' <<< "$FILES" && ! grep -qE '^src/' <<< "$FILES"; then
+              RUN=false
+            fi
+          fi
+          echo "Full build and test matrix: $RUN"
+          echo "run-build-tests=$RUN" >> "$GITHUB_OUTPUT"
+
       - name: Checkout code
         if: ${{ steps.skip-check.outputs.skip == 'false' }}
         uses: actions/checkout@v6
@@ -86,7 +106,9 @@ jobs:
   cache-sources:
     name: Cache depends sources
     needs: [check-skip]
-    if: ${{ needs.check-skip.outputs.skip == 'false' }}
+    if: |
+      needs.check-skip.outputs.skip == 'false' &&
+      needs.check-skip.outputs.run-build-tests == 'true'
     uses: ./.github/workflows/cache-depends-sources.yml
     with:
       runs-on: ${{ needs.check-skip.outputs['runner-arm64'] }}
@@ -94,7 +116,9 @@ jobs:
   container:
     name: Build container
     needs: [check-skip]
-    if: ${{ needs.check-skip.outputs.skip == 'false' }}
+    if: |
+      needs.check-skip.outputs.skip == 'false' &&
+      needs.check-skip.outputs.run-build-tests == 'true'
     uses: ./.github/workflows/build-container.yml
     with:
       context: ./contrib/containers/ci
-- 
```
?


This `classify_ci_changes.py` is over-complicated

File: .github/workflows/test_classify_ci_changes.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+SPEC = importlib.util.spec_from_file_location("classify_ci_changes", MODULE_PATH)
+MODULE = importlib.util.module_from_spec(SPEC)
+assert SPEC.loader is not None
+SPEC.loader.exec_module(MODULE)
```
Comment (thepastaclaw) [inline .github/workflows/test_classify_ci_changes.py:12]: **🟡 Suggestion: Run the classifier regression suite automatically**

The new regression suite is only executed manually in the PR's testing instructions. Repository CI runs `ci/dash/lint.sh`, which invokes `test/lint/all-lint.py`; that runner discovers only `test/lint/lint-*.py`, and no workflow or other CI script invokes this module. Future changes to this CI-critical classifier can therefore pass CI without exercising the exclusion-boundary, fail-open, rename, malformed-input, or summary-escaping tests. Wire this inexpensive suite into the always-running lint/check path.

<sub>source: ['claude']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Sol-only technical fallback

The classifier is conservative at collection and packaging boundaries, and its 13-test regression suite passes locally. One in-scope test-coverage gap remains: no repository CI target executes the new suite, so future classifier regressions can merge without exercising it.
Source: reviewers gpt-5.6-sol and glm-5.3-flash; final verifier gpt-5.6-sol.

One or more required Phase-1 GLM Flash lanes remained technically unusable after the bounded exact-model retry. Their evidence was discarded as authoritative, and the complete selected role cohort was rerun fresh on exact gpt-5.6-sol before this fresh Sol verifier produced the final decision. No additional Phase-2 reviewer pass ran.

### Review provenance
- Phase 1 GLM evidence: **technically unusable after bounded retry; discarded from the decision**
- GLM failure attempts: `codex-dash-core-commit-history-54e142283faf47e8804d5bc330a42e7f` (completed), `codex-general-7328c0e9802445e898fe36c7e8f9eb99` (failed), `codex-general-5acba548b2f04eec87bee0bd239c8456` (failed)
- Sol-only fallback reasons: `launch_transport_or_nonzero_exit`
- Sol-only fallback reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Fresh verifier (Sol): `gpt-5.6-sol` — final-verifier
- Additional Phase 2 pass: **not run; the Sol-only fallback is final**

🟡 1 suggestion(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `.github/workflows/test_classify_ci_changes.py`:
- [SUGGESTION] .github/workflows/test_classify_ci_changes.py:8-12: Run the classifier regression suite automatically
  The new regression suite is only executed manually in the PR's testing instructions. Repository CI runs `ci/dash/lint.sh`, which invokes `test/lint/all-lint.py`; that runner discovers only `test/lint/lint-*.py`, and no workflow or other CI script invokes this module. Future changes to this CI-critical classifier can therefore pass CI without exercising the exclusion-boundary, fail-open, rename, malformed-input, or summary-escaping tests. Wire this inexpensive suite into the always-running lint/check path.
```
</details>
